pep-0655.rst

```diff
@@ -121,6 +120,96 @@ items where it is redundant, to enable additional explicitness if desired:
        title: Required[str]  # redundant
        year: NotRequired[int]
 
+It is an error to use both ``Required[]`` and ``NotRequired[]`` at the
+same time:
+
+::
+
+   class Movie(TypedDict):
+       title: str
+       year: NotRequired[Required[int]]  # ERROR
+
+
+The `alternative syntax <https://www.python.org/dev/peps/pep-0589/#alternative-syntax>`__
+for TypedDict also supports
+``Required[]`` and ``NotRequired[]``:
+
+::
+
+   Movie = TypedDict('Movie', {'name': str, 'year': NotRequired[int]})
+
+
+Interaction with ``Annotated[]``
+-----------------------------------
+
+``Required[]`` and ``NotRequired[]`` can be used with ``Annotated[]``,
+in any nesting order:
+
+::
+
+   class Movie(TypedDict):
+       title: str
+       year: NotRequired[Annotated[int, ValueRange(-9999, 9999)]]  # ok
+
+::
+
+   class Movie(TypedDict):
+       title: str
+       year: Annotated[NotRequired[int], ValueRange(-9999, 9999)]  # ok
+
+
+Interaction with ``get_type_hints()``
+-------------------------------------
+
+``typing.get_type_hints(...)`` applied to a TypedDict will by default
+strip out any ``Required[]`` or ``NotRequired[]`` type qualifiers,
+since these qualifiers are expected to be inconvenient for code
+casually introspecting type annotations.
+
+``typing.get_type_hints(..., include_extras=True)`` however
+*will* retain ``Required[]`` and ``NotRequired[]`` type qualifiers,
+for advanced code introspecting type annotations that
+wishes to preserve *all* annotations in the original source:
+
+::
+
+   class Movie(TypedDict):
+       title: str
+       year: NotRequired[int]
+   
+   assert get_type_hints(Movie) == \
+       {'title': str, 'year': int}
+   assert get_type_hints(Movie, include_extras=True) == \
+       {'title': str, 'year': NotRequired[int]}
+
+
+Interaction with ``get_origin()`` and ``get_args()``
+----------------------------------------------------
+
+``typing.get_origin()`` and ``typing.get_args()`` will be updated to
+recognize ``Required[]`` and ``NotRequired[]``:
+
+::
+
+   assert get_origin(Required[int]) is Required
+   assert get_args(Required[int]) == (int,)
+   
+   assert get_origin(NotRequired[int]) is NotRequired
+   assert get_args(NotRequired[int]) == (int,)
+
+
+Interaction with ``__required_keys__`` and ``__optional_keys__``
+----------------------------------------------------------------
+
+An item marked with ``Required[]`` will always appear
+in the ``__required_keys__`` for its enclosing TypedDict. Similarly an item
+marked with ``NotRequired[]`` will always appear in ``__optional_keys__``.
+
+::
+
+   assert Movie.__required_keys__ == frozenset({'title'})
```

Comment:
fyi, the typing-extensions implementation doesn't do this. Implementing this behavior would require the TypedDict constructor to inspect the annotations on each value, which is somewhat fragile and won't work if you use typing.TypedDict with typing_extensions.Required.

As the author of a tool that needs to understand this at runtime, I'd be OK with changing the spec so that Required and NotRequired aren't reflected in __required_keys__ and __optional_keys__.

Reply:
> Implementing this behavior would require the TypedDict constructor to inspect the annotations on each value

Aye.

> won't work if you use typing.TypedDict with typing_extensions.Required.

typing_extensions.Required is identical to typing.Required at runtime if the latter is available:
```
# typing_extensions.py
if hasattr(typing, 'Required'):
    Required = typing.Required
    NotRequired = typing.NotRequired
```

> is somewhat fragile

Doesn't seem to me like the code to do the inspection would be very complex: It just has to unwrap `Annotated[]` values until it hits `(Not)Required[]` or something else. And then toss out the `(Not)Required[]` wrapper before stuffing the result into the appropiate `__*_keys__` attribute.

Reply:
> typing_extensions.Required is identical to typing.Required at runtime if the latter is available:

But on 3.8 through 3.10 typing.TypedDict exists and doesn't know about typing_extensions.Required.

Reply:
> As the author of a tool that needs to understand this at runtime, I'd be OK with changing the spec so that Required and NotRequired aren't reflected in **required_keys** and **optional_keys**.

👍

Reply:
> But on 3.8 through 3.10 typing.TypedDict exists and doesn't know about typing_extensions.Required.

Good point: I probably can't update typing.TypedDict in 3.8-3.10 to support a feature from 3.11.

Looks like the strategy in the past for this type of issue is to make typing_extensions.TypedDict recognize *all* supported syntax in the latest accepted typing PEPs, with users of typing.TypedDict sometimes receiving incomplete functionality. For example, under that interpretation:

* typing.TypedDict in Python 3.8-3.10 would not understand typing_extensions.Required and would put something like `typing_extensions.Required[T]` incorrectly into `MyTypingDict.__optional_keys__` rather than putting `T` into `MyTypingDict.__required_keys__`.
* typing_extensions.TypedDict in Python 3.8-3.10 *would* correctly understand typing_extensions.Required.
* typing_extensions.TypedDict in Python 3.11+ would alias typing.TypedDict, which would correctly recognize both typing.Required and typing_extensions.Required (because the latter would alias the former).

Does that sound reasonable @JelleZijlstra ?

(If so I should update the "How to Teach This" section to recommend use of typing_extensions.TypedDict over typing.TypedDict in Python <= 3.10 when used in conjunction with typing_extensions.Required. 🎗️ )

Reply:
That would work and it's fine with me.

I'd still be OK with just skipping this work completely. My experience is that runtime checks in typing.py are fragile (in case a later PEP changes what kind of things are accepted at runtime) and often don't do exactly what I'd want when consuming the annotations. For example, in pyanalyze I'd still want to parse Required/NotRequired myself for use cases where users use 3.8-3.10 typing.TypedDict. Therefore, I'd generally favor minimizing runtime work and just making sure whatever the user wrote is recoverable through runtime introspection.